### ISSUE TITLE
Memory management in the storr API

### ISSUE THREAD

**Issue by wlandau** [richfitz/storr#69]
From https://github.com/richfitz/storr/issues/55#issuecomment-344623527 and https://github.com/richfitz/storr/issues/55#issuecomment-367837436: in ropensci/drake#155, @kendonB noticed that `rescue_cache()` is hogging memory, a problem I thought I had solved. Does `storr` keep in memory objects loaded with `$get()`? I am currently using:

```r
rescue_cache <- function(
  targets = NULL,
  path = getwd(), search = TRUE, verbose = TRUE, force = FALSE,
  cache = drake::get_cache(
    path = path, search = search, verbose = verbose, force = force),
  jobs = 1
){
  if (is.null(cache)){
    return(invisible())
  }
  for (namespace in cache$list_namespaces()){
    X <- cache$list(namespace = namespace)
    if (!is.null(targets)){
      X <- intersect(X, targets)
    }
    lightly_parallelize(
      X = X,
      FUN = rescue_del,
      jobs = jobs,
      cache = cache,
      namespace = namespace
    )
  }
  invisible(cache)
}

rescue_del <- function(key, cache, namespace){
  tryCatch(
    cache$get(key = key, namespace = namespace),
    error = function(e){
      cache$del(key = key, namespace = namespace)
    }
  )
  invisible()
}
```

with some supporting functions:

```r
lightly_parallelize <- function(X, FUN, jobs = 1, ...) {
  jobs <- safe_jobs(jobs)
  if (is.atomic(X)){
    lightly_parallelize_atomic(X = X, FUN = FUN, jobs = jobs, ...)
  } else {
    mclapply(X = X, FUN = FUN, mc.cores = jobs, ...)
  }
}

lightly_parallelize_atomic <- function(X, FUN, jobs = 1, ...){
  jobs <- safe_jobs(jobs)
  keys <- unique(X)
  index <- match(X, keys)
  values <- mclapply(X = keys, FUN = FUN, mc.cores = jobs, ...)
  values[index]
}

safe_jobs <- function(jobs){
  ifelse(on_windows(), 1, jobs)
}

on_windows <- function(){
  this_os() == "windows"
}
```

**Comment by richfitz** [comment]
Thanks for reposting this.

My gut feeling is that this is a documentation issue, so can you

(1) rerun your code that was memory intensive with and without changing this line

```r
    cache$get(key = key, namespace = namespace),
```

to

```r
    cache$get(key = key, namespace = namespace, use_cache = FALSE),
```

and monitor the memory usage to confirm that this makes the problem go away.  If not, it's a storr bug but I don't think that's the case here.

Which leads to the second bug: why is this unclear?  I see I have not mentioned it in the main storr vignette and I should add something there.  This is documented [here](http://richfitz.github.io/storr/reference/storr.html#methods) for the `get` method as:

> `use_cache`: Use the internal cache to avoid reading or writing to the underlying storage if the data has already been seen (i.e., we have seen the hash of the object before).

Can you tell me if you (a) saw this and figured that it was irrelevant or unclear, or (b) missed it?

**Comment by wlandau** [comment]
I did see the `use_cache` argument, but I thought it was unclear. Both the name and the help file leave me wondering what else we would use if we weren't using the cache. But as soon as you remind me that the alternative is to keep stuff in memory, I get it immediately.

**Comment by richfitz** [comment]
OK, cool. I'll have a go at clarifying the docs for the next release. Does it help the example you posted?

**Comment by wlandau** [comment]
Thanks, Rich!

I am not actually sure `use_cache` will help with the supposed memory issue in that example. `use_cache` is already `TRUE` by default, so I am not sure anymore where the memory issues are coming from.

**Comment by richfitz** [comment]
Ah, this is very useful.  The documentation should make it clear that what you want is the opposite!  You want to set `use_cache = FALSE`.  Reading what is there I can see that this is not properly discussed.

What happens is this: when you use `get`, storr reads the id of the hash for your key.  Then it reads the data from wherever that is stored.  The difference is that we always have to re-read the hash given the key, but the data is stored by content so if we have already read the data corresponding to a given hash we can avoid reading it again.  Because both deserialisation and i/o are relatively expensive this is a good thing to do most of the time.

What `use_cache = TRUE` does is after storr reads the data, it keeps a copy in an in-memory cache.  It's just an R environment with a key being the hash and the value being the object.   When we ask for the data given the hash we first check the internal cache and only read from disk.

Because of the way that R's environments are stored it's not very straightforward to order the cache so that only the last 'n' entries are kept etc (that could be done but I'm not sure it could be done efficiently enough to be worth it). So it grows without bounds.  There is a `flush_cache` method that will remove all objects from the cache.

So in your case, because you're going to iterate over all data in the cache, and because you really want to detect if the objects are ok, you really want `use_cache = FALSE`

Does that make sense?

**Comment by wlandau** [comment]
Yeah, that makes sense. I think the problem is that the term "cache" is ambiguous. You're caching with memory in one sense and with the file system in another. Do you plan to change the argument name to something like `use_memory`?

**Comment by richfitz** [comment]
No sorry - that would be a breaking change and I don't think it's a better name.  I'll improve the docs though.

Does changing the value help with your memory use?

**Comment by kendonB** [comment]
I haven't noticed any problems with memory for a few months

>

**Comment by wlandau** [comment]
Fair point. I reverted some insertions of `use_cache = FALSE` but kept ones where I think the data could be large. We'll see how performance goes.

**Comment by kendonB** [comment]
Is there a feasible solution that lets the R garbage collector handle this? As in the objects in the memory cache would remain there and be reusable until garbage collection happens?

**Comment by wlandau** [comment]
By the way, does `storr` cap its memory usage from `use_cache = TRUE`? Related: https://github.com/ropensci/drake/pull/302#issuecomment-370921125. cc @krlmlr

**Comment by richfitz** [comment]
No, it does not limit cache size - doing that would require significant additional work (and therefore need careful thought before using) because we'd need to order the cache as well to ensure that it behaved in a fifo way.

I'd be happy to entertain a PR for this, or other improvements to the caching layer, but in general this is not something that I think that it should be that necessary to think about.  It's only going to affect people where you have lots of distinct bits of data floating around.  Other options would be to limit the size of objects that come into the cache, but even that is going to require an additional `object.size` call per `get` which is going to have an overhead on it

**Comment by wlandau** [comment]
A good place to start might be to respect the memory limits of the R session itself. Beyond that, decisions get trickier, I agree.

**Comment by wlandau** [comment]
I am picturing a situation where `storr` consumes so much memory that R starts producing errors like "cannot allocate a vector of size 1 Gb"

**Comment by richfitz** [comment]
That's quite different to not respecting "the memory limits of the R session itself" - it's just what happens when you're out of memory.

There is no straightforward data structure that gets you everything you want here.  Either you turn off caching, in which case every large object will be read from disk, and deserialised, every time, which will be very slow, or we store every object that the storr touches in cache.

It might be possible to replace the environment with (say) a list so that there is an ordering, or keep an ordering vector there, and store sizes in order to keep the total size down but will require a new object type and will need some careful work to ensure that it's not much slower (accessing a classed object has an unfortunately high overhead - as seen in the [R6 performance vignettes](https://cran.r-project.org/web/packages/R6/vignettes/Performance.html) and [cache invalidation](https://en.wikipedia.org/wiki/Cache_invalidation) is one of [the two hard problems in computer science](https://martinfowler.com/bliki/TwoHardThings.html).

**Comment by krlmlr** [comment]
Maybe *storr* can purge a fraction of its cache upon request, oldest first, ignoring object size?

**Comment by richfitz** [comment]
There is a `flush_cache` method for pruning the whole cache at the moment.  Any more than that would require an additional data structure for recording when the data was added, or ordering the data within the structure (I'm happy to be proved wrong, and happy for data structures that make the caching later suit this use case better).

**Comment by krlmlr** [comment]
We could still purge objects at random, feels better than "all or nothing". And then talk about it:

```r
flush_cache(0.3)
#> Removed 333 out of 1110 objects in the cache, recovered 12 GB
```

A `gc()` in the end would be useful for end users, too.

We could also remove those objects that [`!MAYBE_SHARED()`](https://cran.r-project.org/doc/manuals/r-devel/R-exts.html#Some-convenience-functions) first, with or without sampling, but this may require some C code.

**Comment by richfitz** [comment]
I don't really see the advantage of a random purge - that just adds nondeterministic behaviour.  And cleaning the cache out is always "safe" - it just may slow things down a bit.

I'm wondering if we can use weak references in the cache to leave the objects subject to garbage collection though.  That definitely requires C code, but that or your `MAYBE_SHARED` approach might help with getting the right objects

**Comment by wlandau** [comment]
In light of https://github.com/ropensci/drake/issues/930#issuecomment-508616847, I would like to revisit this issue. I am interested in reducing the memory footprint of `$set(use_cache = FALSE)`.
